File: sdk/basin-php/src/Basin/Storage/StorageClient.php
<?php

declare(strict_types=1);

namespace Basin\Storage;

use Basin\Exception\BasinApiException;
use Basin\Http\Transport;
use Basin\Types\Bucket;
use Basin\Types\DownloadResult;
use Basin\Types\SignedUrl;
use Basin\Types\StorageObject;

/**
 * Storage client — bindings for /storage/v1/*.
 *
 * Route sources (verified, crates/basin-rest/src/server.rs):
 *   POST   /storage/v1/bucket                              create bucket
 *   GET    /storage/v1/bucket/:name                        bucket metadata
 *   DELETE /storage/v1/bucket/:name                        delete + purge
 *   POST   /storage/v1/object/:bucket/*path                upload (JWT)
 *   GET    /storage/v1/object/:bucket/*path                download (JWT)
 *   DELETE /storage/v1/object/:bucket/*path                delete (JWT)
 *   POST   /storage/v1/object/list/:bucket                 list (body: prefix/limit/offset)
 *   DELETE /storage/v1/object/:bucket                      bulk delete (body: prefixes[])
 *   GET    /storage/v1/object/public/:project_id/:bucket/*path  public download (no JWT)
 *   POST   /storage/v1/object/sign/upload/:bucket/*path    mint signed URL (JWT)
 *   GET    /storage/v1/object/sign/:project/:bucket/*path  signed download (no JWT)
 *
 * Usage:
 *   $client->storage()->createBucket('avatars', public: true);
 *   $client->storage()->fromBucket('avatars')->upload('user.jpg', $bytes, 'image/jpeg');
 */
class StorageClient
{
    public function __construct(
        private readonly Transport $transport,
        private readonly \Closure $projectIdGetter,
    ) {
    }

    // ------------------------------------------------------------------
    // Bucket operations
    // ------------------------------------------------------------------

    /**
     * POST /storage/v1/bucket (201).
     *
     * @param string[] $allowedMimeTypes
     */
    public function createBucket(
        string $name,
        bool $public = false,
        ?int $fileSizeLimit = null,
        array $allowedMimeTypes = [],
    ): Bucket {
        $data = (array) $this->transport->requestJson('POST', '/storage/v1/bucket', body: [
            'name'               => $name,
            'public'             => $public,
            'file_size_limit'    => $fileSizeLimit,
            'allowed_mime_types' => $allowedMimeTypes,
        ]);
        return Bucket::fromArray($data);
    }

    /**
     * GET /storage/v1/bucket/:name.
     */
    public function getBucket(string $name): Bucket
    {
        $data = (array) $this->transport->requestJson('GET', '/storage/v1/bucket/' . rawurlencode($name));
        return Bucket::fromArray($data);
    }

    /**
     * DELETE /storage/v1/bucket/:name — also purges object bytes.
     * @return array<string,mixed>
     */
    public function deleteBucket(string $name): array
    {
        return (array) $this->transport->requestJson('DELETE', '/storage/v1/bucket/' . rawurlencode($name));
    }

    /**
     * Return an object-level client scoped to one bucket.
     */
    public function fromBucket(string $bucket): StorageBucketClient
    {
        return new StorageBucketClient($this->transport, $bucket, $this->projectIdGetter);
    }
}

/**
 * Object-level operations scoped to one bucket.
 */
class StorageBucketClient
{
    public function __construct(
        private readonly Transport $transport,
        private readonly string $bucket,
        private readonly \Closure $projectIdGetter,
    ) {
    }

    // ------------------------------------------------------------------
    // Object operations
    // ------------------------------------------------------------------

    /**
     * POST /storage/v1/object/:bucket/*path — upload a file.
     */
    public function upload(string $path, string $data, ?string $contentType = null): StorageObject
    {
        $headers = [];
        if ($contentType !== null) {
            $headers['Content-Type'] = $contentType;
        }

        $result = (array) $this->transport->requestJson(
            'POST',
            $this->objectPath($path),
            body: $data,
            headers: $headers,
        );
        return StorageObject::fromArray($result);
    }

    /**
     * GET /storage/v1/object/:bucket/*path — authenticated download.
     */
    public function download(string $path): DownloadResult
    {
        $response = $this->transport->raw('GET', $this->objectPath($path));
        return new DownloadResult(
            data: (string) $response->getBody(),
            contentType: $response->getHeaderLine('Content-Type') ?: null,
            etag: $response->getHeaderLine('ETag') ?: null,
        );
    }

    /**
     * DELETE /storage/v1/object/:bucket/*path — single delete.
     * @return array<string,mixed>
     */
    public function remove(string $path): array
    {
        return (array) $this->transport->requestJson('DELETE', $this->objectPath($path));
    }

    /**
     * POST /storage/v1/object/list/:bucket.
     * @return StorageObject[]
     */
    public function list(?string $prefix = null, ?int $limit = null, ?int $offset = null): array
    {
        $data = $this->transport->requestJson(
            'POST',
            '/storage/v1/object/list/' . rawurlencode($this->bucket),
            body: ['prefix' => $prefix, 'limit' => $limit, 'offset' => $offset],
        );
        return array_map(
            fn(array $d) => StorageObject::fromArray($d),
            (array) ($data ?? []),
        );
    }

    /**
     * DELETE /storage/v1/object/:bucket with {prefixes:[...]} — bulk delete.
     *
     * RLS-denied objects are skipped server-side.
     *
     * @param string[] $prefixes
     * @return array<string,mixed>
     */
    public function removeByPrefixes(array $prefixes): array
    {
        return (array) $this->transport->requestJson(
            'DELETE',
            '/storage/v1/object/' . rawurlencode($this->bucket),
            body: ['prefixes' => $prefixes],
        );
    }

    /**
     * Build the no-JWT public URL.
     *
     * GET /storage/v1/object/public/:project_id/:bucket/*path
     * Only works for buckets with public=true.
     */
    public function getPublicUrl(string $path): string
    {
        $project = ($this->projectIdGetter)();
        if ($project === null) {
            throw new BasinApiException(
                'E_INVALID_REQUEST',
                'public URLs require a project id (pass project_id to new Client())',
                0,
            );
        }
        $base = $this->transport->getBaseUrl();
        return "{$base}/storage/v1/object/public/{$project}/{$this->encodedBucket()}/{$this->encodePath($path)}";
    }

    /**
     * POST /storage/v1/object/sign/upload/:bucket/*path — mint signed download URL.
     *
     * TTL is capped at 7 days server-side; default 3600 s.
     * Despite the 'upload' literal in the path this mints a download URL —
     * that segment only disambiguates the axum router (storage_sign.rs).
     */
    public function createSignedUrl(string $path, int $expiresIn = 3600): SignedUrl
    {
        $data = (array) $this->transport->requestJson(
            'POST',
            "/storage/v1/object/sign/upload/{$this->encodedBucket()}/{$this->encodePath($path)}",
            body: ['expires_in' => $expiresIn],
        );
        return SignedUrl::fromArray($data, $this->transport->getBaseUrl());
    }

    // ------------------------------------------------------------------
    // Internal helpers
    // ------------------------------------------------------------------

    private function objectPath(string $path): string
    {
        return "/storage/v1/object/{$this->encodedBucket()}/{$this->encodePath($path)}";
    }

    private function encodedBucket(): string
    {
        return rawurlencode($this->bucket);
    }

    private function encodePath(string $path): string
    {
        return implode('/', array_map('rawurlencode', explode('/', $path)));
    }
}

File: sdk/basin-java/src/main/java/io/basin/sdk/StorageClient.java
package io.basin.sdk;

import com.fasterxml.jackson.databind.JsonNode;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;

/**
 * Storage client for {@code /storage/v1/*} routes.
 *
 * <p>Route sources (verified against {@code crates/basin-rest/src/server.rs:373-424};
 * handlers in {@code crates/basin-rest/src/routes/storage.rs} and
 * {@code storage_sign.rs}):
 * <ul>
 *   <li>{@code POST   /storage/v1/bucket}                                  create bucket
 *   <li>{@code GET    /storage/v1/bucket/:name}                            bucket metadata
 *   <li>{@code DELETE /storage/v1/bucket/:name}                            delete + purge
 *   <li>{@code POST   /storage/v1/object/:bucket/*path}                    upload (JWT)
 *   <li>{@code GET    /storage/v1/object/:bucket/*path}                    download (JWT)
 *   <li>{@code DELETE /storage/v1/object/:bucket/*path}                    delete (JWT)
 *   <li>{@code POST   /storage/v1/object/list/:bucket}                     list objects
 *   <li>{@code DELETE /storage/v1/object/:bucket}                          bulk delete
 *   <li>{@code GET    /storage/v1/object/public/:project_id/:bucket/*path} public download
 *   <li>{@code POST   /storage/v1/object/sign/upload/:bucket/*path}        mint signed URL
 * </ul>
 *
 * <p>Obtain via {@link BasinClient#storage()}.
 */
public final class StorageClient {

    private final BasinTransport transport;
    private final Supplier<String> projectIdSupplier;

    StorageClient(BasinTransport transport, Supplier<String> projectIdSupplier) {
        this.transport = transport;
        this.projectIdSupplier = projectIdSupplier;
    }

    // ------------------------------------------------------------------
    // Bucket operations
    // ------------------------------------------------------------------

    /**
     * {@code POST /storage/v1/bucket} (201) — create a bucket.
     *
     * @param name             bucket name
     * @param isPublic         whether objects are publicly accessible without a token
     * @param fileSizeLimit    maximum object size in bytes; {@code null} for no limit
     * @param allowedMimeTypes mime type allowlist; {@code null} or empty for all types
     */
    public CompletableFuture<StorageBucket> createBucket(String name, boolean isPublic,
                                                          Long fileSizeLimit,
                                                          List<String> allowedMimeTypes) {
        Map<String, Object> body = new HashMap<>();
        body.put("name", name);
        body.put("public", isPublic);
        if (fileSizeLimit != null) body.put("file_size_limit", fileSizeLimit);
        body.put("allowed_mime_types", allowedMimeTypes != null ? allowedMimeTypes : List.of());
        return transport.executeJsonAsync(transport.jsonPost("/storage/v1/bucket", null, body, true))
                .thenApply(n -> transport.readValue(n, StorageBucket.class));
    }

    /** Blocking variant of {@link #createBucket}. */
    public StorageBucket createBucketBlocking(String name, boolean isPublic,
                                               Long fileSizeLimit, List<String> allowedMimeTypes) {
        return createBucket(name, isPublic, fileSizeLimit, allowedMimeTypes).join();
    }

    /** {@code GET /storage/v1/bucket/:name} — bucket metadata. */
    public CompletableFuture<StorageBucket> getBucket(String name) {
        return transport.executeJsonAsync(
                transport.jsonGet("/storage/v1/bucket/" + encodePath(name), null, true))
                .thenApply(n -> transport.readValue(n, StorageBucket.class));
    }

    /** Blocking variant of {@link #getBucket}. */
    public StorageBucket getBucketBlocking(String name) {
        return getBucket(name).join();
    }

    /** {@code DELETE /storage/v1/bucket/:name} — delete bucket and purge all objects. */
    public CompletableFuture<JsonNode> deleteBucket(String name) {
        return transport.executeJsonAsync(
                transport.jsonDelete("/storage/v1/bucket/" + encodePath(name), null, null, true));
    }

    /** Blocking variant of {@link #deleteBucket}. */
    public JsonNode deleteBucketBlocking(String name) {
        return deleteBucket(name).join();
    }

    /**
     * Return an object-level client scoped to one bucket.
     *
     * <pre>{@code
     * StorageBucketClient bucket = client.storage().fromBucket("avatars");
     * bucket.upload("users/alice.png", bytes, "image/png").join();
     * }</pre>
     */
    public StorageBucketClient fromBucket(String bucket) {
        return new StorageBucketClient(transport, bucket, projectIdSupplier);
    }

    // ------------------------------------------------------------------
    // Path encoding
    // ------------------------------------------------------------------

    static String encodePath(String segment) {
        return URLEncoder.encode(segment, StandardCharsets.UTF_8).replace("+", "%20");
    }

    static String encodeObjectPath(String path) {
        String[] parts = path.split("/", -1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < parts.length; i++) {
            if (i > 0) sb.append('/');
            sb.append(encodePath(parts[i]));
        }
        return sb.toString();
    }
}
